File: src/mb/mizinkobusters/kitpvp/kit/Counter.java
package mb.mizinkobusters.kitpvp.kit;

import mb.mizinkobusters.kitpvp.utils.KitPvPUtils;
import mb.mizinkobusters.kitpvp.utils.PlayerSalvationUtils;
import org.bukkit.Sound;
import org.bukkit.entity.Arrow;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.player.PlayerRespawnEvent;

import java.util.Random;

public class Counter implements Listener {

    @EventHandler
    public void onKill(PlayerRespawnEvent event) {
        Player player = event.getPlayer();
        if (player.getKiller() == null) {
            return;
        }

        Player killer = player.getKiller();
        if (!KitPvPUtils.isInWorld(killer)) {
            return;
        }
        if (!KitPvPUtils.getKit(killer).equals("Counter")) {
            return;
        }
        PlayerSalvationUtils.heal(player);
    }

    @EventHandler
    public void onDamage(EntityDamageByEntityEvent event) {
        if (!event.getEntity().getType().equals(EntityType.PLAYER)) {
            return;
        }

        Player damagee = (Player) event.getEntity();
        if (!KitPvPUtils.isInWorld(damagee)) {
            return;
        }
        if (!KitPvPUtils.getKit(damagee).equals("Counter")) {
            return;
        }

        if (!(event.getDamager().getType().equals(EntityType.PLAYER)
                || event.getDamager().getType().equals(EntityType.ARROW))) {
            return;
        }

        Player damager = null;
        if (event.getDamager().getType().equals(EntityType.PLAYER)) {
            damager = (Player) event.getDamager();
        }
        if (event.getDamager().getType().equals(EntityType.ARROW)) {
            Arrow arrow = (Arrow) event.getDamager();
            damager = (Player) arrow.getShooter();
        }

        Random r = new Random();
        int i = r.nextInt(4);
        if (i == 0) {
            damager.damage(event.getDamage() / 2, damagee);
            event.setDamage(event.getDamage() / 2);
            damagee.playSound(damagee.getLocation(), Sound.BLOCK_ANVIL_HIT, 1, 0);
            damagee.sendMessage("§7Skill Trigger: §aCounter");
        }
    }
}

File: src/main/java/mb/mizinkobusters/kitpvp/kit/Counter.kt
package mb.mizinkobusters.kitpvp.kit

import java.util.Random
import org.bukkit.Material
import org.bukkit.Sound
import org.bukkit.entity.Arrow
import org.bukkit.entity.Player
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.inventory.ItemStack

object Counter : BaseKit {
    override val kitTypeInfo = "§aGuard"
    override val isBillingKit = true
    override val lore = listOf(
        "§f頭: 鉄",
        "§f胴: ダイヤ",
        "§f腰: チェーン",
        "§f脚: 鉄",
        "§f武器: 石の剣",
        "§f補助: 金のリンゴ×1",
        "§f特殊: 1/5の確率で被ダメージを半減し,",
        "§f被ダメージを相手に反射する"
    )
    override val displayItemStack = ItemStack(Material.DIAMOND_CHESTPLATE)

    override val helmet = ItemStack(Material.IRON_HELMET)
    override val chestplate = ItemStack(Material.DIAMOND_CHESTPLATE)
    override val leggings = ItemStack(Material.CHAINMAIL_LEGGINGS)
    override val boots = ItemStack(Material.IRON_BOOTS)

    override val weapons = listOf(
        ItemStack(Material.STONE_SWORD),
        ItemStack(Material.GOLDEN_APPLE)
    )

    override fun onDamageByEntity(event: EntityDamageByEntityEvent) {
        super.onDamageByEntity(event)

        val entity = event.entity
        if (entity !is Player) return

        val damager = event.damager
        if (damager !is Arrow) return
        if (damager.shooter !is Player) return

        val r = Random()
        val i = r.nextInt(4)
        if (i == 0) {
            (damager.shooter as Player).damage(event.damage / 2, entity)
            event.damage = event.damage / 2
            entity.playSound(entity.location, Sound.BLOCK_ANVIL_HIT, 1f, 0f)
            entity.sendMessage("§7Skill Trigger: §aCounter")
        }
    }
}